constified implementations of `Default`

fee1-dead opened the pull request:
set `issue = "none"` for now before deciding if we can have `impl const`s in the standard library after #86750 and #86775 have landed.

rust-highfive commented:
r? @scottmcm

(rust-highfive has picked a reviewer for you, use r? to override)

jyn514 commented on library/alloc/src/string.rs:
```diff
@@ -2106,7 +2106,8 @@ impl_eq! { Cow<'a, str>, &'b str }
 
 #[stable(feature = "rust1", since = "1.0.0")]
-impl Default for String {
+#[rustc_const_unstable(feature = "const_default_impls", issue = "none")]
```
Are these shown in rustdoc? Not sure if there's much point in adding these without, it would be very hard to find.

fee1-dead replied:
These attributes are required to gate the const impl. I will look at how rustdoc deals with this and possibly open a separate PR

fee1-dead replied:
Also the RFC for `impl const` is not even accepted, so I don't know if we should change rustdoc for this. What do you think @jyn514? [tracking issue](67792)

> This RFC has not yet been accepted. It is being implemented on a provisional basis to [evaluate the potential fallout](https://github.com/rust-lang/rfcs/pull/2632#issuecomment-567699174).

jyn514 replied:
If the RFC hasn't been accepted, I definitely don't think we should add it to the standard library.

CDirkx commented:
What would be the policy for changing to `const impl`s in the standard library? Is that something we want to do very conservatively, only to test the implementation for the RFC, or more liberally. I know of quite some const code that would become a lot more readable when able to use `const impl Eq` etc., but given that the RFC hasn't been accepted yet I don't know how much the library team wants to commit to this.

scottmcm commented on library/alloc/src/lib.rs:
```diff
@@ -69,6 +69,7 @@
 #![warn(missing_docs)]
 #![warn(missing_debug_implementations)]
 #![allow(explicit_outlives_requirements)]
+#![allow(incomplete_features)]
```
I'm torn here.  I would of course love to have these be `impl const`, but if it needs `incomplete_features` it feels like maybe core/alloc/std shouldn't be using it yet.

Maybe @m-ou-se would have thoughts here?

Mark-Simulacrum replied:
I agree that this seems premature. The latest updates from lang team backlog bonanzas (as of several months ago) seem to suggest we are not yet committed to a specific behavior/path of implementation, so exposing (even unstably) const trait impls seems like it's not a good idea quite yet.

fee1-dead commented:
Something for the lang team: what would make the feature *complete*? 

I have opened PRs for all issues listed on the [tracking issue](https://github.com/rust-lang/rust/issues/67792) that blocked having `impl const`s in the standard library (#86750 #86775 #86571), so what are the next steps?

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3082816950) [(plain)](https://github.com/rust-lang/rust/commit/df5541c21c4d2dde8d8b83482334dd179b218bff/checks/3082816950/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling libc v0.2.93
   Compiling std v0.0.0 (/checkout/library/std)
   Compiling compiler_builtins v0.1.47
   Compiling unwind v0.0.0 (/checkout/library/unwind)
error: the feature `const_trait_impl` is incomplete and may not be safe to use and/or cause compiler crashes
    |
109 | #![feature(const_trait_impl)]
    |            ^^^^^^^^^^^^^^^^
    |
    |
    = note: `-D incomplete-features` implied by `-D warnings`

error: aborting due to previous error

error: could not compile `core`
```

</details>

fee1-dead commented:
Update: in [this zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/146212-t-compiler.2Fconst-eval/topic/implementation.20path.20for.20const.20trait.20impls) we felt comfortable marking the feature not incomplete and start experimenting (non generic impls) in libstd.

Ping @scottmcm @Mark-Simulacrum

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3082885936) [(plain)](https://github.com/rust-lang/rust/commit/c841090195205f82ab8812ab4dbd562d26e3c897/checks/3082885936/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling libc v0.2.93
   Compiling std v0.0.0 (/checkout/library/std)
   Compiling compiler_builtins v0.1.47
   Compiling unwind v0.0.0 (/checkout/library/unwind)
error: the feature `const_trait_impl` is incomplete and may not be safe to use and/or cause compiler crashes
    |
109 | #![feature(const_trait_impl)]
    |            ^^^^^^^^^^^^^^^^
    |
    |
    = note: `-D incomplete-features` implied by `-D warnings`

error: aborting due to previous error

error: could not compile `core`
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3187820232) [(plain)](https://github.com/rust-lang/rust/commit/29c0954762a16474e33ea9143259940405b52de6/checks/3187820232/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling libc v0.2.98
   Compiling std v0.0.0 (/checkout/library/std)
   Compiling compiler_builtins v0.1.47
   Compiling unwind v0.0.0 (/checkout/library/unwind)
error: the feature `const_trait_impl` is incomplete and may not be safe to use and/or cause compiler crashes
    |
113 | #![feature(const_trait_impl)]
    |            ^^^^^^^^^^^^^^^^
    |
    |
    = note: `-D incomplete-features` implied by `-D warnings`

error: aborting due to previous error

error: could not compile `core`
```

</details>

Mark-Simulacrum commented:
> Update: in [this zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/146212-t-compiler.2Fconst-eval/topic/.E2.9C.94.20implementation.20path.20for.20const.20trait.20impls) we felt comfortable marking the feature not incomplete and start experimenting (non generic impls) in libstd.

Have the incomplete / not yet ready for use parts of the feature gate been split out yet? I'm not sure we should be using it until that has been done: we've run into ongoing trouble with specialization, for example, precisely because of the actually-not-ok bits in min_specialization. I'd prefer to avoid doing so with const trait if there are known limitations.

Regardless, I think a decision from T-libs-api on what the policy for const-unstable trait impls should look like would be merited. I'm a little surprised this is *possible* since historically impls weren't possible to feature gate - is the story different in the const case for some reason?

fee1-dead commented:
> > Update: in [this zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/146212-t-compiler.2Fconst-eval/topic/.E2.9C.94.20implementation.20path.20for.20const.20trait.20impls) we felt comfortable marking the feature not incomplete and start experimenting (non generic impls) in libstd.
> 
> Have the incomplete / not yet ready for use parts of the feature gate been split out yet? I'm not sure we should be using it until that has been done: we've run into ongoing trouble with specialization, for example, precisely because of the actually-not-ok bits in min_specialization. I'd prefer to avoid doing so with const trait if there are known limitations.

I don't think we currently have incomplete / not yet ready for use parts. There are actually aspects of the RFC that are yet to be implemented, but I don't think this PR can expose not-ok stuff to users.

> I'm a little surprised this is _possible_ since historically impls weren't possible to feature gate - is the story different in the const case for some reason?

AFAIK the stability should be inherited and _maybe_ this is only true for const stability. But I did add a test for this: https://github.com/rust-lang/rust/blob/master/src/test/ui/rfc-2632-const-trait-impl/staged-api.rs

oli-obk commented:
> is the story different in the const case for some reason?

Yes it is. const stability is a separate system that does not participate in trait resolution, so you don't get type-system unsoundness where a trait is both implemented and not. Instead you just get an error if you call a non-const impl from a const context.

fee1-dead commented:
Ping @rust-lang/libs-api

dtolnay reviewed (commented):
> I think a decision from T-libs-api on what the policy for const-unstable trait impls should look like would be merited.

I don't think we have any objection to starting to land const impls as long as they are properly gated, i.e. stable code cannot be broken by ongoing churn in the const impl feature, all the way through its stabilization or eventual removal. I can't judge that from the PR but if the compiler team asserts it's properly gated, then it's ready.

As for which impls to make const, I think the bar is going to be the same as for const inherent methods: it has to be reasonably usable in const or have an extremely strong justification otherwise. For example we regularly reject const on an inherent method of a type that is impossible to construct in const, even if that method's implementation is otherwise compatible with const.

fee1-dead commented:
I think:
 - I am sure that non-generic const impls will be good; (gated behind `#[rustc_const_unstable]` of course :) ) 
 - a bit less sure about generic const fns (should be better after #87375)
 - not so sure about implementing traits with associated types
 - Don't think we should constify impls with trait bounds now, because of potential regressions (see [this](https://github.com/rust-lang/rust/pull/86853#issuecomment-891036254)) I need to add a few ui tests tracking this.

bors commented:
:umbrella: The latest upstream changes (presumably #87725) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

yaahc commented:
>     * I am sure that non-generic const impls will be good;

When you say "non-generic" in this case you mean traits that don't have generic parameters like `Default`, and not impls that don't have generic parameters like `impl<T> Default for Vec<T>` correct? Just trying to gauge if you're indicating that these `Default` impls should be good to go or not.

fee1-dead commented:
Sorry, to be clear, I meant that const impls without trait bounds other than `Sized` (and also `: ?Sized` bound) will be good, so yes, I think these `Default` implementations are good to go.

oli-obk commented:
the library changes lgtm.

I would like to see a few ui tests before merging this. I know we test the feature itself, and we test the stability things themselves, but now that libstd has such const impls, we should also test a few of them, as well as their successful gating.

The main PR message could also reflect the discussion and its result.

bors commented:
:umbrella: The latest upstream changes (presumably #87768) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

fee1-dead commented:
I have tried to add a ui test but it keeps failing. Perhaps we need to wait for #87375 to land first.

@rustbot label -S-waiting-on-author S-blocked

fee1-dead committed:
Constified `Default` implementations

The libs-api team agrees to allow const_trait_impl to appear in the
standard library as long as stable code cannot be broken (they are
properly gated) this means if the compiler teams thinks it's okay, then
it's okay.

My priority on constifying would be:

	1. Non-generic impls (e.g. Default) or generic impls with no
	   bounds
	2. Generic functions with bounds (that use const impls)
	3. Generic impls with bounds
	4. Impls for traits with associated types

For people opening constification PRs: please cc me and/or oli-obk.

fee1-dead commented:
@rustbot label -S-blocked +S-waiting-on-author

r? @oli-obk

oli-obk commented:
> I would like to see a few ui tests before merging this. I know we test the feature itself, and we test the stability things themselves, but now that libstd has such const impls, we should also test a few of them, as well as their successful gating.

I would still sleep more soundly having a ui test showing that using these without feature gates causes an error

fee1-dead committed:
Add ui test

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit c5983c8111658f851cc3b8fff909322d255239af has been approved by `oli-obk`

<!-- @bors r=oli-obk c5983c8111658f851cc3b8fff909322d255239af -->
<!-- homu: {"type":"Approved","sha":"c5983c8111658f851cc3b8fff909322d255239af","approver":"oli-obk"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #88094) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 6bd2ecba726f06587d7d60ffca5605841ac39091 has been approved by `oli-obk`

<!-- @bors r=oli-obk 6bd2ecba726f06587d7d60ffca5605841ac39091 -->
<!-- homu: {"type":"Approved","sha":"6bd2ecba726f06587d7d60ffca5605841ac39091","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit 6bd2ecba726f06587d7d60ffca5605841ac39091 with merge 1805ef4dd4df410428fb1cd37a47f88781bc469c...
<!-- homu: {"type":"BuildStarted","head_sha":"6bd2ecba726f06587d7d60ffca5605841ac39091","merge_sha":"1805ef4dd4df410428fb1cd37a47f88781bc469c"} -->

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/3351337078) [(plain)](https://github.com/rust-lang-ci/rust/commit/1805ef4dd4df410428fb1cd37a47f88781bc469c/checks/3351337078/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

oli-obk commented:
@bors retry

bors commented:
:hourglass: Testing commit 6bd2ecba726f06587d7d60ffca5605841ac39091 with merge adf168844791b213383b18d10515eaee88695f7d...
<!-- homu: {"type":"BuildStarted","head_sha":"6bd2ecba726f06587d7d60ffca5605841ac39091","merge_sha":"adf168844791b213383b18d10515eaee88695f7d"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3356505426)
Approved by: oli-obk
Pushing adf168844791b213383b18d10515eaee88695f7d to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3356505426"},"merge_sha":"adf168844791b213383b18d10515eaee88695f7d"} -->
